Caltech-IPAC/firefly
# IRSA-504 workspace server api

I've checked in the manager for workspaces that should work for different implementation. Part of the code was already implemented few years ago so i needed to adjusted and add/fix webdav specific implementation.

**The idea** is to have a common interface that will be used from UI client to interact with workspaces back in the server. Right now the implementation is WebDAV and local FS (not fully implemented) for testing purposes. I've kept original method that was used in other places already (such as to get the ws directory or staging files directly from FF app) - _`I'm not sure if we want to stage files directly to workspace without the consent from the user first, such that the user has to deliberately set the default mechanism to save the result into workspaces`_

**The API** exposes `get, put, create parent, rename, delete and list files` from user workspace independently of the implementation in the workspace server. There is also a way to set and get metadata, which will help in particular to get/set custom properties in the future. 

I've added a bunch of tests that covers the implementation of the manager and aslo helps to understand how the api works. Typically a request to the server will give back a metadata object which contains the properties of the node(s). This object should be also generic to cover any workspace implementation in the future. 

**The tests** covers the authenticated user and public (anonymous) space as well. I've omitted the test to authenticate with cookies (with IRSA SSO server).

Finally, i've added some code to start the **skeleton of the server commands** but i feel that this will be part of the other ticket IRSA-474 (File picker to use server worskspace api). 

Couple of things need to be addressed and **can't be done at this stage**: 

- how the response json object would look like when used by the file picker to list the files for example.
- Should we create by default the workspace user home if not present? Right now the implementation is creating a parent (see constructor of the manager).
- Would the file picker only show files if logged in? Should we always use anonymous space?

**Need to discuss further also:** I'm not sure if we want to stage files directly to workspace without the consent from the user first, such that the user has to deliberalty set the default mechanism to save the result into workspaces

Part of the _**answer of many questions could be found in IRSA-457**_: workspace reqs and concepts, when ready.

> **To TEST**:
> 
> Checkout branches from firefly and firefly_test_data: `IRSA-504-workspace-server-api`.
> In order to see logs, you need to change level from WARN to INFO in `config/log4j-test.properties`.
> Then run tests `gradle test` or from your favorite IDE. Tests will take files from `test_data` here: `edu/caltech/ipac/firefly/ws/`. One fits file is used for SSO only because pubspace has limit to 1Mb. The tests use a common user id: `test-ws` with no password (anonymous).
> 
> Then go to some tests and see how you can call the manager via a handler that will keep the manager mapping the user id. Each handler belongs to a protocol: WEBDAV, LOCAL or VOSPACE. Manager can be of different implementation: webdav or local.
> 
> Another way to see in action is to comment out the tear down (@After in test) of the test and look for url:  https://irsa.ipac.caltech.edu/pubspace/test-ws/
> This area should be populated and clean up as long you run those tests.

Let me know if you have trouble or find problems.
